java/memory/src/main/java/io/netty/buffer/ArrowBuf.java

```diff
@@ -499,461 +225,212 @@ public boolean equals(Object obj) {
   }
 
-  /**
-   * @see ArrowBuf#setLongLE(int, long).
-   */
-  @Override
-  public void _setLongLE(int index, long value) {
-    setLongLE(index, value);
-  }
+  // TODO: do bound checking properly for the following APIs
 
-  @Override
-  public ArrowBuf getBytes(int index, ByteBuf dst, int dstIndex, int length) {
-    udle.getBytes(index + offset, dst, dstIndex, length);
+  public ArrowBuf getBytes(int index, byte[] dst, int dstIndex, int length) {
+    final int end = dstIndex + length - 1;
+    for (int i = dstIndex; i <= end; i++) {
+      dst[i] = PlatformDependent.getByte(addr(index));
+    }
     return this;
   }
 
-  @Override
-  public ArrowBuf getBytes(int index, OutputStream out, int length) throws IOException {
-    udle.getBytes(index + offset, out, length);
+  public ArrowBuf getBytes(int index, ByteBuffer dst) {
+    int toRead = length;
+    long address = addr(index);
+    // word by word memcpy
+    while (toRead >= 8) {
+      dst.putLong(PlatformDependent.getLong(address));
+      address += 8;
+      toRead -= 8;
+    }
+    while (toRead > 0) {
+      dst.put(PlatformDependent.getByte(address));
+      --toRead;
+    }
     return this;
   }
 
-  @Override
-  public int getBytes(int index, GatheringByteChannel out, int length) throws IOException {
-    return udle.getBytes(index + offset, out, length);
-  }
-
-  @Override
-  public int getBytes(int index, FileChannel out, long position, int length) throws IOException {
-    return udle.getBytes(index + offset, out, position, length);
+  public ArrowBuf setBytes(int index, byte[] src, int srcIndex, int length) {
+    if (length != 0) {
+      // memcpy from source byte array at the starting index (srcIndex)
+      // to the underlying chunk of memory this ArrowBuf has access to
+      // at the given index (index)
+      PlatformDependent.copyMemory(src, srcIndex, addr(index), (long)length);
+    }
+    return this;
   }
 
-  @Override
-  public ArrowBuf setBytes(int index, ByteBuf src, int srcIndex, int length) {
-    udle.setBytes(index + offset, src, srcIndex, length);
+  public ArrowBuf setBytes(int index, ByteBuffer src) {
+    int length = src.remaining();
+    long dstAddress = addr(index);
+    if(length != 0) {
+      if(src.isDirect()) {
+        final long srcAddress = PlatformDependent.directBufferAddress(src) + (long)src.position();
+        PlatformDependent.copyMemory(srcAddress, dstAddress, (long)src.remaining());
+        src.position(src.position() + length);
+      } else if(src.hasArray()) {
+        PlatformDependent.copyMemory(src.array(), src.arrayOffset() + src.position(), dstAddress, (long)length);
+        src.position(src.position() + length);
+      } else {
+        while (length >= 8) {
+          PlatformDependent.putLong(dstAddress, src.getLong());
+          length -= 8;
+          dstAddress += 8;
+        }
+        while (length > 0) {
+          PlatformDependent.putByte(dstAddress, src.get());
```

Comment:
Do you need to decrement `length` and increment `dstAddress`?

Reply:
should be fixed in latest commit